vowpalwabbit/interaction_ground.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "reductions.h"
+#include "label_dictionary.h"
+#include "vw.h"
+#include "vw_exception.h"
+#include "shared_data.h"
+#include "label_parser.h"
+#include "cb_adf.h"
+#include "cb_algs.h"
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+using namespace CB_ALGS;
+namespace Interaction_Ground
+{
+struct interaction_ground
+{
+  double total_importance_weighted_reward;//the accumulated importance weighted reward of a policy which optimizes the given value
+  double total_uniform_reward;
+  double total_importance_weighted_cost;//the accumulated importance weighted loss of the policy which optimizes the negative of the given value.
+  double total_uniform_cost;
+public:
+  interaction_ground() { total_importance_weighted_reward = 0.; total_uniform_reward = 0.; total_importance_weighted_cost = 0.; total_uniform_cost = 0.;};
+};
+
+  void negate_cost(multi_ex& ec_seq)
+  {
+    for (auto* example_ptr : ec_seq)
+      for (auto& label : example_ptr->l.cb.costs)
+        if (label.has_observed_cost())
+          label.cost = -label.cost;
+  }
+
+void learn(interaction_ground& ig, multi_learner& base, multi_ex& ec_seq)
+{
+  //find reward of sequence
+  CB::cb_class label = CB_ADF::get_observed_cost_or_default_cb_adf(ec_seq);
+  ig.total_uniform_cost += label.cost / label.probability / ec_seq.size();//=p(uniform) * IPS estimate
+  ig.total_uniform_reward += -label.cost / label.probability / ec_seq.size();
+
+  //find prediction & update for cost
+  base.predict(ec_seq);
+  ig.total_importance_weighted_cost += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
+  base.learn(ec_seq);
+
+  //find prediction & update for reward
+  label.cost = -label.cost;
+  base.predict(ec_seq,1);
+  ig.total_importance_weighted_reward += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
+
+  //change cost to reward
+  negate_cost(ec_seq);
+  base.learn(ec_seq,1);
+  negate_cost(ec_seq);
+}
+
+void predict(interaction_ground& ig, multi_learner& base, multi_ex& ec_seq)
+{
+  //figure out which is better by our current estimate.
+  if (ig.total_uniform_cost - ig.total_importance_weighted_cost
+      > ig.total_uniform_reward - ig.total_importance_weighted_reward)
+    base.predict(ec_seq);
+  else
+    base.predict(ec_seq,1);
+}
+
+}  // namespace Interaction_Ground
+using namespace Interaction_Ground;
+base_learner* interaction_ground_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  vw& all = *stack_builder.get_all_pointer();
```

Comment:
looks like line 75 isn't needed since `all` variable isn't being used anywhere (that is why the CI is failing here)